### PR TITLE
feat: Added JSON configuration.

### LINKED ISSUE #4: config
Allow users to add their own links with a KV file or similar, because why not I suppose

### PR DESCRIPTION
**Pull request by Trico-Everfire**
Added JSON Configuration, feel free to tear it apart. it's a pretty basic implementation.

resolves #4

### CODE REVIEW

File: src/config.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+				MediaItem_t item = section.items.at( j );
+
+				QJsonObject appItems1;
+				appItems1["name"] = item.name;
+				appItems1["icon"] = item.icon;
+				appItems1["urlType"] = item.urlType;
+				appItems1["url"] = item.url;
+				QJsonArray arguments1;
+
+				for(int k = 0; k < item.args.count(); k++){
```
Comment (AWildErin) [inline src/config.h]: [fromStringList](https://doc.qt.io/qt-5/qjsonarray.html#fromStringList)

File: src/main.cpp [hunk 2]
```diff
@@ -16,6 +22,15 @@ int main(int argc, char** argv)
 	QString styleSheet = QLatin1String( file.readAll() );
 	qApp->setStyleSheet( styleSheet );
 
+	QFile configFile("./config.json");
+	if(!configFile.exists()){
```
Comment (AWildErin) [inline src/main.cpp]: Formatting. Also see mainview.cpp for a simillar comment on this

File: src/mainview.cpp [hunk 3]
```diff
@@ -94,6 +73,7 @@ void CMainView::OpenUrl( QString url )
 void CMainView::OpenProcess( QString execName, QStringList params )
 {
 	auto pProcess = new QProcess( this );
-	
-	pProcess->start( execName, params);
```
Comment (AWildErin) [inline src/mainview.cpp:98]: Why did you change this?

File: src/mainview.cpp [hunk 4]
```diff
@@ -1,81 +1,60 @@
...
+	QFile config( "./config.json" );
+	config.open( QFile::ReadOnly );
+	QJsonDocument JSONConfigDocument = QJsonDocument::fromJson( config.readAll() );
+	config.close();
+	QJsonObject JSONConfig = JSONConfigDocument.object();
+
+	for ( auto it = JSONConfig.begin(); it != JSONConfig.end(); it++ )
 	{
-		MediaSection_t section = Sections.at( i );
+		// MediaSection_t section = Sections.at( it.key() );
```
Comment (AWildErin) [inline src/mainview.cpp]: Remove the comment

File: src/mainview.cpp [hunk 5]
```diff
@@ -1,81 +1,60 @@
...
 
-CMainView::CMainView( QWidget *pParent ) : QDialog( pParent )
+CMainView::CMainView( QWidget *pParent ) :
+	QDialog( pParent )
 {
 	auto pLayout = new QVBoxLayout( this );
 	pLayout->setObjectName( "SDKLayout" );
 
-	for ( int i = 0; i < Sections.count(); i++ )
+	QFile config( "./config.json" );
```
Comment (AWildErin) [inline src/mainview.cpp]: Couldn't we ultimately just load the default config in here rather than in `main.cpp`? Also this config path isn't really ideal. This would only work if `config.json` is next to the launcher.

File: src/config.h [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+			"Documentation",
+			{
+				{ "Valve Developer Community", ":/resource/vdc.png", "url", "https://developer.valvesoftware.com", { "" } },
+				{ "Chaos Wiki", ":/resource/chaos.png", "url", "https://chaosinitiative.github.io/Wiki/", { "" } },
+				{ "Momentum Wiki", ":/resource/momentum.png", "url", "https://docs.momentum-mod.org/", { "" } }
+			}
+		}
+	});
+
+ QJsonDocument defaultConfig(){
```
Comment (AWildErin) [inline src/config.h]: Formatting

File: src/config.h [hunk 7]
```diff
@@ -0,0 +1,85 @@
+//
```
Comment (AWildErin) [inline src/config.h]: Kind of a useless comment here, we don't really use this sort of comment

File: src/main.cpp [hunk 8]
```diff
@@ -1,9 +1,15 @@
...
 #include <QIcon>
 #include <QFile>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QJsonArray>
+#include "config.h"
 
 #include "mainview.h"
 
+QJsonDocument defaultConfig();
```
Comment (AWildErin) [inline src/main.cpp]: nit: Why is this declared here when we also declare this in the `config.h`?

**Review by AWildErin — CHANGES_REQUESTED** [review]

File: src/mainview.cpp (hunk 3 above)
Comment (Trico-Everfire) [reply src/mainview.cpp:98]: I was debugging a problem, and ended up changing it slightly, I can revert it back if you want.

Comment (AWildErin) [reply src/mainview.cpp:98]: Please do, it's just less code honestly

File: src/main.cpp (hunk 8 above)
Comment (Trico-Everfire) [reply src/main.cpp]: Accidental leftover, will be removed.

File: src/mainview.cpp (hunk 5 above)
Comment (Trico-Everfire) [reply src/mainview.cpp]: How would you have it? there is no real way to "search" for the config. it has to be stored somewhere.

File: src/config.h (hunk 7 above)
Comment (Trico-Everfire) [reply src/config.h]: a CMake automatic generated comment. will be removed.

File: src/mainview.cpp (hunk 5 above)
Comment (Trico-Everfire) [reply src/mainview.cpp]: Would you like for me to move it to `mainview.cpp` and then have it use QDir::currentDir()? or how would you resolve the file's location?

Comment (AWildErin) [reply src/mainview.cpp]: Just don't worry about it for now. I'll work on that when it comes to creating a project system for source mods.

Comment (Trico-Everfire) [reply src/mainview.cpp]: All right.

File: src/config.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+			MediaItem_t item = section.items.at( j );
+
+			QJsonObject appItems1;
+			appItems1["name"] = item.name;
+			appItems1["icon"] = item.icon;
+			appItems1["urlType"] = item.urlType;
+			appItems1["url"] = item.url;
+			QJsonArray arguments1;
+
+			appItems1["args"] = QJsonArray::fromStringList(item.args);
```
Comment (AWildErin) [inline src/config.h]: nit: formatting

File: src/main.cpp [hunk 10]
```diff
@@ -16,6 +18,15 @@ int main(int argc, char** argv)
 	QString styleSheet = QLatin1String( file.readAll() );
 	qApp->setStyleSheet( styleSheet );
 
+	QFile configFile( "./config.json" );
+	if ( !configFile.exists() )
```
Comment (AWildErin) [inline src/main.cpp]: couldn't we maybe move this to mainview.cpp? Not really required but I don't see a point to re-implementing it here when we also try to access the json file in mainview.cpp

File: src/mainview.cpp [hunk 11]
```diff
@@ -94,6 +71,5 @@ void CMainView::OpenUrl( QString url )
 void CMainView::OpenProcess( QString execName, QStringList params )
 {
 	auto pProcess = new QProcess( this );
-	
-	pProcess->start( execName, params);
+	pProcess->start(execName,params);
```
Comment (AWildErin) [inline src/mainview.cpp]: nit: formatting

**Review by AWildErin — COMMENTED** [review]
Some more smaller things that ideally I'd like to be looked at

**Review by AWildErin — APPROVED** [review]
